File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        left, right = 0, len(height) - 1
        maxL, maxR = height[left], height[right]
        res = 0
        while left < right:
            if maxL <= maxR:
                left += 1
                elevation = maxL - height[left]
                res += elevation if elevation > 0 else 0
                maxL = max(maxL, height[left]) 
            else:
                right -= 1
                elevation = maxR - height[right]
                res += elevation if elevation > 0 else 0
                maxR = max(maxR, height[right])
        return res

File: Data Structures & Algorithms/palindrome-partitioning/submission-1.cpp
class Solution {
public:
    
    vector<vector<string>> res;
    vector<string> path;

    bool isPalindrome(int left, int right, const string& s) {
        while (left < right) {
            if (s[left] != s[right]) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    void backtrack(const string& s, int start) {
        if (start == s.size()) {
            res.push_back(path);
            return;
        }

        for (int end = start; end < s.size(); end++) {
            if (isPalindrome(start, end, s)) {
                path.push_back(s.substr(start, end - start + 1));
                backtrack(s, end + 1);
                path.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        backtrack(s, 0);
        return res;
    }
};
